dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+
+from funcy import cached_property
+from ratelimit import limits, sleep_and_retry
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.cached_dirs = {}
+        self.cached_ids = {}
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        if Config.SECTION_REMOTE_KEY_FILE not in self.config:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at https://dvc.org/doc."
+            )
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE
+        )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    @sleep_and_retry
```

Comment:
do we need all three decorators here? I'm curios why.

Reply:
I believe, API querying should always work with an assumption to exceed API usage limits.
 `@limits(calls=10, period=1)` - accepts 10 calls per 1 second and raise custom exception if limit is exceeded.
`@sleep_and_retry` - catches custom limit exceeded exceptions raised by `limits` decorator and put thread to sleep and retry the call.
`@on_exception` - should handle all still occurred exceptions to allow us work on max possible frequency of API queries. If I'm not wrong, setting limits to 8 calls per 2 seconds helps fully to avoid usage limits exceeded exceptions, but we want max possible speed and stability with random API failures. ( for the moment still needed to adjust exceptions catching and raising code to increase stability )

Reply:
I understand `@on_exception` with an exponential backoff, but `limit` + `sleep_and_retry` seem excessive and I'm not sure they can work in multi-user environment for example. Would it better to just have `@on_exception` and let's it figure out how much it needs to wait to avoid hitting the limits?

Reply:
seems `@on_exception` works great, thanks!